### PR TITLE
feat:`locations.geojson` file parsing and geometry validation

### PR DESCRIPTION
**Pull request by cka-y**
**Summary:**
Part of #1850
This PR introduces enhancements for validating GeoJSON files. The most notable changes include the addition of new validation notice classes, the implementation of a `GeoJsonGeometryValidator` class, and the renaming of classes to align with a consistent naming convention.

### Validation Notices:
* **New Notice Classes**: 
  * `InvalidGeometryNotice` - Triggered when a GeoJSON file contains invalid geometry that does not adhere to the specified structure.
  * `MissingRequiredElementNotice` - Raised when elements required in the GeoJSON format are absent.
  * `UnsupportedGeometryTypeNotice` - Issued for geometry types that are not currently supported by our validator (supporting only `Polygon` and `MultiPolygon`).

### Dependency Updates:
* **New Dependency**: 
  * Added `org.locationtech.jts:jts-core:1.20.0` to manage geometric operations in compliance with the GeoJSON standards. The update is reflected in `main/build.gradle`.

### Geometry Validation

* According to the OpenGIS specification, which follows the Open Geospatial Consortium's Simple Feature Specification (OGC SFS), a valid polygon must :
   * Consist of closed, simple linear rings that do not self-intersect.
   * Have an exterior ring defining the boundary of the polygon.
   * Contain zero or more interior rings (holes) that are entirely within the exterior ring.
   * Ensure that rings do not cross each other.
* The validation logic utilizes the `isValid` function from the `locationtech.jts` library to verify that polygons conform to these criteria, as outlined by the OGC SFS specification, which is an implementation of the OpenGIS standards for geographic information systems. [Reference documentation](https://javadoc.io/doc/org.locationtech.jts/jts-core/1.18.1/org/locationtech/jts/geom/Geometry.html#isValid--).

**Expected Behavior:** 
To verify these changes, you can use the provided [test feed example](https://github.com/user-attachments/files/17297482/Archive.zip). Below is a sample validation result, showing expected error notices:

Example Validation Results

Other notices covered in this PR:
- Unsupported GeoJSON `type`
image
- Malformed JSON file
image
- Unsupported feature type
image

Example of problem location from [mdb-2052](https://mobilitydatabase.org/feeds/mdb-2052) feature `area_915` (self intersection in the red square):
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJSONFileLoader.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+      JsonArray featuresArray = jsonObject.getAsJsonArray("features");
+      int i = 0;
+      for (JsonElement feature : featuresArray) {
+        GtfsGeoJSONFeature gtfsGeoJSONFeature = extractFeature(feature, noticeContainer, i++);
+        hasUnparsableFeature |= gtfsGeoJSONFeature == null;
+        if (gtfsGeoJSONFeature != null) {
+          features.add(gtfsGeoJSONFeature);
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJSONFileLoader.java]: done ✅

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java [hunk 2]
```diff
@@ -605,11 +605,16 @@ public ImmutableSortedSet<String> getFilenames() {
...
+      return specFeatures.keySet().stream()
+          .anyMatch(
+              feature ->
+                  feature.getFeatureGroup() != null
+                      && feature.getFeatureGroup().equals("Flexible Services")
+                      && specFeatures.get(feature));
+
+    } catch (Exception e) {
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java]: done ✅

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJSONFileLoader.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  private static void addMissingRequiredFieldsNotice(
+      List<String> missingRequiredFields,
+      NoticeContainer noticeContainer,
+      String featureId,
+      int featureIndex) {
+    if (featureId == null) {
+      featureId = "N/A";
+    }
```
[the remark this answers is not in the document]
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJSONFileLoader.java]: done ✅

**Comment by emmambd** [comment]
As discussed with @tzujenchanmbd, a partial list of changes: 

1) (Not sure if this is a change) Modify `missing_required_element` so `properties` must be included. It can be empty, but if the `properties` object is not included, the notice should trigger. 
2) Modify `unsupported_location_type` to `unsupported_geojson_type`

**Comment by tzujenchanmbd** [comment]
Not sure if it's all included in this PR, just try to clarify everything regarding `type`.

There 3-levels `type` in the spec. 
type `FeatureCollection`
type `Feature`
type  `Polygon` `MultiPolygon`

Perhaps there could be 3 different notice name for each:
unsupported_geojson_type
unsupported_features_type
unsupported_geometry_type (already used in screenshot)

**Comment by cka-y** [comment]
@tzujenchanmbd @emmambd Could you specify what would be the columns/description for :
- unsupported_geojson_type
- unsupported_features_type

Could you also confirm that the names and descriptions for the existing notices are ok? (except unsupported_location_type that will be modified)

**Comment by tzujenchanmbd** [comment]
Notice name, column displayed, and description lgtm from the screenshot.

**Comment by emmambd** [comment]
@cka-y Approved notice details: 

**unsupported_geojson_type**
Description:  An unsupported geojson type is used in the locations.geojson file. Use FeatureCollection instead to comply with the spec.
Columns: geojsonType

**unsupported_feature_type**
Description:  An unsupported feature type is used in the locations.geojson file. Use Feature instead to comply with the spec.
Columns: featureIndex, featureId, featureType

Once added, should be good on a QA side! Interested for the memory usage acceptance test to be run on this now that they're merged.

**Comment by cka-y** [comment]
@emmambd done ✅ 
only thing is that, in order to respect the naming convention of the classes, the `unsupported_geojson_type` is actually `unsupported_geo_json_type`.

**Comment by cka-y** [comment]
# Visual Justification of Differences in Acceptance Tests

The acceptance tests reveal a new `invalid_geometry` notice for the following feeds:
- [mdb-2052](https://mobilitydatabase.org/feeds/mdb-2052)
- [mdb-2043](https://mobilitydatabase.org/feeds/mdb-2043)

### For `mdb-2052`:
The issue is caused by the features `area_915` and `area_917`.

- **area_915**: The following images illustrate the `Polygon` geometry. The self-intersecting loop in the polygon clearly violates the geometry specifications. The problematic area is highlighted in the red square:

  <div style="display: flex; justify-content: space-between;">
    Area 915 - Overview
    Area 915 - Zoomed
  </div>

- **area_917**:  
  <div style="display: flex; justify-content: space-between;">
    Area 917 - Overview
    Area 917 - Zoomed
  </div>

### For `mdb-2043`:
The issue is with the feature `area_921`:

  <div style="display: flex; justify-content: space-between;">
    Area 921 - Overview
    Area 921 - Zoomed
  </div>

**Comment by jcpitre** [comment]
Do you know why for some files git know it's a rename (e.g. GtfsGeojsonFeaturesContainer.java to GtfsGeoJsonFeaturesContainer.java) and others not (e.g. GeojsonFileLoader.java to GeoJsonFileLoader.java). 
It makes it hard to follow the changes.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+      ValidatorProvider validatorProvider,
+      InputStream inputStream,
+      NoticeContainer noticeContainer) {
+    GtfsGeoJsonFileDescriptor geoJsonFileDescriptor = (GtfsGeoJsonFileDescriptor) fileDescriptor;
+    geometryValidator = new GeoJsonGeometryValidator(noticeContainer);
+    try {
+      List<GtfsGeoJsonFeature> entities = extractFeaturesFromStream(inputStream, noticeContainer);
+      return geoJsonFileDescriptor.createContainerForEntities(entities, noticeContainer);
+    } catch (JsonParseException jpex) {
+      noticeContainer.addValidationNotice(new MalformedJsonNotice(GtfsGeoJsonFeature.FILENAME));
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:43]: There's no useful information that we can extract from the JsonParseException and put in the notice?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      logger.atSevere().withCause(jpex).log("Malformed JSON in locations.geojson");
+      return fileDescriptor.createContainerForInvalidStatus(TableStatus.UNPARSABLE_ROWS);
+    } catch (IOException ioex) {
+      noticeContainer.addSystemError(new IOError(ioex));
+      return fileDescriptor.createContainerForInvalidStatus(TableStatus.UNPARSABLE_ROWS);
+    } catch (UnparsableGeoJsonFeatureException ugex) {
+      logger.atSevere().withCause(ugex).log("Unparsable GeoJSON feature");
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:50]: Is there a reason why we generate a notice for JsonParseException but not for UnparsableGeoJsonFeatureException?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java (hunk 4 above)
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:43]: Not sure -- could open an issue to investigate the different errors as this is already large

Comment (jcpitre) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:43]: Yeah, it's already good that we have what you added, the rest is just icing on the cake.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+            GtfsGeoJsonFeature.FEATURE_COLLECTION_FIELD_NAME
+                + '.'
+                + GtfsGeoJsonFeature.FEATURE_TYPE_FIELD_NAME);
+      } else if (!featureObject
+          .get(GtfsGeoJsonFeature.FEATURE_TYPE_FIELD_NAME)
+          .getAsString()
+          .equals("Feature")) {
+        noticeContainer.addValidationNotice(
+            new UnsupportedFeatureTypeNotice(
+                featureIndex,
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:124]: Should this customer facing info be 1 based?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                properties
+                    .get(GtfsGeoJsonFeature.FEATURE_PROPERTIES_STOP_DESC_FIELD_NAME)
+                    .getAsString());
+          }
+
+          return gtfsGeoJsonFeature;
+        }
+      }
+    }
+    addMissingRequiredFieldsNotice(missingRequiredFields, noticeContainer, featureId, featureIndex);
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:210]: Maybe plural, since we potentially add multiple notices
```suggestion
    addMissingRequiredFieldsNotices(missingRequiredFields, noticeContainer, featureId, featureIndex);
```

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    return null;
+  }
+
+  private static void addMissingRequiredFieldsNotice(
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:214]:
```suggestion
  private static void addMissingRequiredFieldsNotices(
```

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java (hunk 6 above)
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:124]: Good point! Thoughts @emmambd ?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidGeometryNotice.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  /** The id of the faulty record. */
+  private final String featureId;
+
+  /** The geometry type of the feature containing the invalid polygon. */
+  private final String geometryType;
+
+  /** The validation error details. */
+  private final String message;
+
+  public InvalidGeometryNotice(String featureId, String geometryType, String validationError) {
```
Comment (jcpitre) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidGeometryNotice.java:41]: Should it have an index like the other notices?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java (hunk 6 above)
Comment (emmambd) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:124]: Isn't common practice in development 0? I'd stick with that, since anyone creating the data in geojson would likely assume the same.

Comment (tzujenchanmbd) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GeoJsonFileLoader.java:124]: yep 0 should be fine.

**Review by jcpitre — APPROVED** [review]
Great work, as usual
